node/src/components/consensus/utils.rs

```diff
@@ -273,4 +300,191 @@ mod tests {
+
+        let insufficient =
+            create_signatures(&mut rng, &validators, INSUFFICIENT_FINALITY_SIGNATURES);
+        let just_enough_weight =
+            create_signatures(&mut rng, &validators, JUST_ENOUGH_FINALITY_SIGNATURES);
+        let too_many = create_signatures(&mut rng, &validators, TOO_MANY_FINALITY_SIGNATURES);
+
+        let result = check_sufficient_finality_signatures(
+            &validator_weights,
+            finality_threshold_fraction,
+            &insufficient,
+        );
+        assert!(matches!(
+            result,
+            Err(FinalitySignatureError::InsufficientWeightForFinality {
+                trusted_validator_weights: _,
+                block_signatures: _,
+                signature_weight,
+                total_validator_weight,
+                finality_threshold_fraction: _
+            }) if *total_validator_weight == TOTAL_VALIDATORS_WEIGHT.into() && *signature_weight == INSUFFICIENT_FINALITY_SIGNATURES.into()
+        ));
+
+        let result = check_sufficient_finality_signatures(
+            &validator_weights,
+            finality_threshold_fraction,
+            &just_enough_weight,
+        );
+        assert!(result.is_ok());
+
+        let result = check_sufficient_finality_signatures(
+            &validator_weights,
+            finality_threshold_fraction,
+            &too_many,
+        );
+        assert!(matches!(
+            result,
+            Err(FinalitySignatureError::TooManySignatures {
+                trusted_validator_weights: _,
+                block_signatures: _,
+                signature_weight,
+                weight_minus_minimum: _,
+                total_validator_weight,
+                finality_threshold_fraction: _
+            }) if *total_validator_weight == TOTAL_VALIDATORS_WEIGHT.into() && *signature_weight == TOO_MANY_FINALITY_SIGNATURES.into()
+        ));
+    }
+
+    #[test]
+    fn finality_signatures_sufficiency_with_quorum_formula() {
+        const TOTAL_VALIDATORS: usize = 20;
+        const TOTAL_VALIDATORS_WEIGHT: usize = TOTAL_VALIDATORS * TEST_VALIDATOR_WEIGHT;
+        const INSUFFICIENT_FINALITY_SIGNATURES: usize = 6;
+        const JUST_ENOUGH_FINALITY_SIGNATURES: usize = 7;
+        const TOO_MANY_FINALITY_SIGNATURES: usize = 8;
+
+        let mut rng = TestRng::new();
+
+        // Total validator weights is 20 (1 for each validator).
+        let (validators, validator_weights) = generate_validators(TOTAL_VALIDATORS_WEIGHT);
+
+        let finality_threshold_fraction = Ratio::new_raw(1, 3);
+
+        // `identity` function is transparent, so the calculated quorum fraction will be equal to
+        // the `finality_threshold_fraction`.
+        let custom_quorum_formula = std::convert::identity;
+
+        // for 20 validators with 20 total validator weight,
+        //   and `finality_threshold_fraction` = 1/3 (~= 6.666)
+        //   and the `quorum fraction` = 1/3         (~= 6.666)
```

Comment:
I think the _quorum fraction_ doesn't convey the meaning. 2/3 is often called _supermajority_, I'm not sure what 1/3 should be called. @afck ?

Reply:
Ah, nvm. I misread the comments I think.